### ISSUE TITLE
Execute zipline shell commands as ipython magic to keep all scripting in a notebook

### ISSUE THREAD

**Issue by jraviotta** [quantopian/zipline#2226]
# Environment

zipline in a docker container

# Description of Issue

Given I am a modern developer with multiple computers running different OS, I want to run all my software in docker containers so that my development environment is portable and consistent.  

Given I am using zipline in a docker container with Jupyter lab, I want to script everything within a notebook for portability. 

Currently, I have to use the command line in the docker container for `ingest` or in a cell with `%%script` magic. This gets pretty cumbersome. It would be much easier to run `%zipline ingest` but that fails.

```
%zipline ingest
UsageErrorTraceback (most recent call last)
<ipython-input-112-984b05ef838d> in <module>()
----> 1 get_ipython().run_line_magic('zipline', 'ingest')

/opt/conda/envs/zipline/lib/python3.5/site-packages/IPython/core/interactiveshell.py in run_line_magic(self, magic_name, line, _stack_depth)
   2129                 kwargs['local_ns'] = sys._getframe(stack_depth).f_locals
   2130             with self.builtin_trap:
-> 2131                 result = fn(*args,**kwargs)
   2132             return result
   2133 

/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/__main__.py in zipline_magic(line, cell)
    292             '%s%%zipline' % ((cell or '') and '%'),
    293             # don't use system exit and propogate errors to the caller
--> 294             standalone_mode=False,
    295         )
    296     except SystemExit as e:

/opt/conda/envs/zipline/lib/python3.5/site-packages/click/core.py in main(self, args, prog_name, complete_var, standalone_mode, **extra)
    694         try:
    695             try:
--> 696                 with self.make_context(prog_name, args, **extra) as ctx:
    697                     rv = self.invoke(ctx)
    698                     if not standalone_mode:

/opt/conda/envs/zipline/lib/python3.5/site-packages/click/core.py in make_context(self, info_name, args, parent, **extra)
    619         ctx = Context(self, info_name=info_name, parent=parent, **extra)
    620         with ctx.scope(cleanup=False):
--> 621             self.parse_args(ctx, args)
    622         return ctx
    623 

/opt/conda/envs/zipline/lib/python3.5/site-packages/click/core.py in parse_args(self, ctx, args)
    883             ctx.fail('Got unexpected extra argument%s (%s)'
    884                      % (len(args) != 1 and 's' or '',
--> 885                         ' '.join(map(make_str, args))))
    886 
    887         ctx.args = args

/opt/conda/envs/zipline/lib/python3.5/site-packages/click/core.py in fail(self, message)
    474         :param message: the error message to fail with.
    475         """
--> 476         raise UsageError(message, self)
    477 
    478     def abort(self):

UsageError: Got unexpected extra argument (ingest)
```

## Here's my hack to get this working as is:
```
import os

# TODO - deal with setting the key in docker
%set_env QUANDL_API_KEY=<my_key>
# Set Zipline data directory to within the project's scope so that I don't pollute my multi-purpose conda container.
os.environ['ZIPLINE_ROOT']=os.path.join(os.getcwd(),'zipline')
# Activate zipline
%load_ext zipline
```
```
%%script bash
/opt/conda/envs/zipline/bin/zipline ingest
```
This works but with the length of time waiting on ingest with no feedback, one starts to wonder if the notebook is hung.

The samples do run...
EG:
```
%%zipline --start 2016-1-1 --end 2018-1-1
from zipline.api import symbol, order, record

def initialize(context):
    pass

def handle_data(context, data):
    order(symbol('AAPL'), 10)
    record(AAPL=data[symbol('AAPL')].price)
```

**Comment by llllllllll** [comment]
If you are already running Python code you can just call the load function yourself.

For example:

```python
from zipline import bundles
bundles.ingest('bundle-name')
```

**Comment by jraviotta** [comment]
Good thought @llllllllll , but no cigar.
from the cli `zipline --help` works.
from a cell in a notebook script magic works as such:
```
%%script bash
/opt/conda/envs/zipline/bin/zipline --help
```
I can import zipline without error as, `import zipline`. 

But from a cell, `from zipline import bundles` fails with:
```
ImportErrorTraceback (most recent call last)
<ipython-input-8-a14f780d0cce> in <module>()
----> 1 from zipline import bundles

ImportError: cannot import name 'bundles'
```
and
```
zipline.bundles.ingest()

AttributeErrorTraceback (most recent call last)
<ipython-input-9-cdb3c6d95b3f> in <module>()
----> 1 zipline.bundles.ingest()

AttributeError: module 'zipline' has no attribute 'bundles'
```
Perhaps something in my setup?

## Environment
```
Linux bf7f37cac929 4.9.87-linuxkit-aufs #1 SMP Wed Mar 14 15:12:16 UTC 2018 x86_64 GNU/Linux
```
```
# packages in environment at /opt/conda:
#
# Name                    Version                   Build  Channel
asn1crypto                0.24.0                   py36_0  
backcall                  0.1.0                      py_0    conda-forge
bleach                    2.1.3                      py_0    conda-forge
ca-certificates           2018.4.16                     0    conda-forge
certifi                   2018.4.16                py36_0    conda-forge
cffi                      1.11.5           py36h9745a5d_0  
chardet                   3.0.4            py36h0f667ec_1  
conda                     4.5.4                    py36_0    conda-forge
conda-env                 2.6.0                h36134e3_1  
cryptography              2.2.2            py36h14c3975_0  
decorator                 4.3.0                      py_0    conda-forge
entrypoints               0.2.3                    py36_1    conda-forge
gmp                       6.1.2                         0    conda-forge
html5lib                  1.0.1                      py_0    conda-forge
idna                      2.6              py36h82fb2a8_1  
ipykernel                 4.8.2                    py36_0    conda-forge
ipython                   6.4.0                    py36_0    conda-forge
ipython_genutils          0.2.0                    py36_0    conda-forge
jedi                      0.12.0                   py36_0    conda-forge
jinja2                    2.10                     py36_0    conda-forge
jsonschema                2.6.0                    py36_1    conda-forge
jupyter_client            5.2.3                    py36_0    conda-forge
jupyter_core              4.4.0                      py_0    conda-forge
jupyterlab                0.32.1                   py36_0    conda-forge
jupyterlab_launcher       0.10.5                   py36_0    conda-forge
libedit                   3.1.20170329         h6b74fdf_2  
libffi                    3.2.1                hd88cf55_4  
libgcc                    7.2.0                h69d50b8_2  
libgcc-ng                 7.2.0                hdf63c60_3  
libsodium                 1.0.16                        0    conda-forge
libstdcxx-ng              7.2.0                hdf63c60_3  
markupsafe                1.0                      py36_0    conda-forge
mistune                   0.8.3                    py36_1    conda-forge
nb_conda                  2.2.1                    py36_0    conda-forge
nb_conda_kernels          2.1.0                    py36_0    conda-forge
nbconvert                 5.3.1                      py_1    conda-forge
nbformat                  4.4.0                    py36_0    conda-forge
nbstripout                0.3.1                    py36_0    conda-forge
ncurses                   6.1                  hf484d3e_0  
nodejs                    9.11.1                        0    conda-forge
notebook                  5.5.0                    py36_0    conda-forge
openssl                   1.0.2o                        0    conda-forge
pandoc                    2.2.1                         0    conda-forge
pandocfilters             1.4.2                    py36_0    conda-forge
parso                     0.2.1                      py_0    conda-forge
pexpect                   4.6.0                    py36_0    conda-forge
pickleshare               0.7.4                    py36_0    conda-forge
pip                       10.0.1                   py36_0  
prompt_toolkit            1.0.15                   py36_0    conda-forge
ptyprocess                0.5.2                    py36_0    conda-forge
pycosat                   0.6.3            py36h0a5515d_0  
pycparser                 2.18             py36hf9f622e_1  
pygments                  2.2.0                    py36_0    conda-forge
pyopenssl                 18.0.0                   py36_0  
pysocks                   1.6.8                    py36_0  
python                    3.6.5                hc3d631a_2  
python-dateutil           2.7.3                      py_0    conda-forge
pyzmq                     17.0.0                   py36_4    conda-forge
readline                  7.0                  ha6073c6_4  
requests                  2.18.4           py36he2e5f8d_1  
ruamel_yaml               0.15.37          py36h14c3975_2  
send2trash                1.5.0                      py_0    conda-forge
setuptools                39.2.0                   py36_0  
simplegeneric             0.8.1                    py36_0    conda-forge
six                       1.11.0           py36h372c433_1  
sqlite                    3.23.1               he433501_0  
terminado                 0.8.1                    py36_0    conda-forge
testpath                  0.3.1                    py36_0    conda-forge
tk                        8.6.7                hc745277_3  
tornado                   5.0.2                    py36_0    conda-forge
traitlets                 4.3.2                    py36_0    conda-forge
urllib3                   1.22             py36hbe7ace6_0  
wcwidth                   0.1.7                    py36_0    conda-forge
webencodings              0.5                      py36_0    conda-forge
wheel                     0.31.1                   py36_0  
xz                        5.2.4                h14c3975_4  
yaml                      0.1.7                had09818_2  
zeromq                    4.2.5                         1    conda-forge
zlib                      1.2.11               ha838bed_2  
```

**Comment by jraviotta** [comment]
Thanks for the clarification. 
This works:
```
from zipline.data import bundles
bundles.ingest('quandl')
```
Interestingly, this does not work: (yes, my api key appears in os.environ)
```
from zipline.data import bundles
bundles.ingest('quandl', show_progress=True)
ValueErrorTraceback (most recent call last)
<ipython-input-3-9c3d5dddc03e> in <module>()
      1 from zipline.data import bundles
----> 2 bundles.ingest('quandl', show_progress=True)

/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/data/bundles/core.py in ingest(name, environ, timestamp, assets_versions, show_progress)
    449                 cache,
    450                 show_progress,
--> 451                 pth.data_path([name, timestr], environ=environ),
    452             )
    453 

/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/data/bundles/quandl.py in quandl_bundle(environ, asset_db_writer, minute_bar_writer, daily_bar_writer, adjustment_writer, calendar, start_session, end_session, cache, show_progress, output_dir)
    207         api_key,
    208         show_progress,
--> 209         environ.get('QUANDL_DOWNLOAD_ATTEMPTS', 5)
    210     )
    211     asset_metadata = gen_asset_metadata(

/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/data/bundles/quandl.py in fetch_data_table(api_key, show_progress, retries)
    111     else:
    112         raise ValueError(
--> 113             "Failed to download Quandl data after %d attempts." % (retries)
    114         )
    115 

ValueError: Failed to download Quandl data after 5 attempts.
```

And
```
from zipline.data import bundles
bundles.ingest( 'quantopian quandl', show_progress=True)

KeyErrorTraceback (most recent call last)
/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/data/bundles/core.py in ingest(name, environ, timestamp, assets_versions, show_progress)
    359         try:
--> 360             bundle = bundles[name]
    361         except KeyError:

KeyError: 'quantopian quandl'

During handling of the above exception, another exception occurred:

UnknownBundleTraceback (most recent call last)
<ipython-input-4-6e8f3e352f62> in <module>()
      1 from zipline.data import bundles
----> 2 bundles.ingest( 'quantopian quandl', show_progress=True)

/opt/conda/envs/zipline/lib/python3.5/site-packages/zipline/data/bundles/core.py in ingest(name, environ, timestamp, assets_versions, show_progress)
    360             bundle = bundles[name]
    361         except KeyError:
--> 362             raise UnknownBundle(name)
    363 
    364         calendar = get_calendar(bundle.calendar_name)

UnknownBundle: No bundle registered with the name 'quantopian quandl'
```

**Comment by llllllllll** [comment]
The first error just means downloading from quandl failed for _some_ reason, but it is not because there is a missing API key. The second error is that the bundle is `quantopian-quandl` with a hyphen, not a space.

**Comment by jraviotta** [comment]
Perhaps this is a bug? In a jupyter notebook I can execute:
```
from zipline.data import bundles
bundles.ingest('quandl')
```
But I get the same `ValueError: Failed to download Quandl data after 5 attempts.` when adding the argument `show_progress=True`.

from bash, I can execute:
```
python
from zipline.data import bundles
bundles.ingest('quandl')
```
So, it appears the problem is asking for progress from a Jupyter notebook.

**Comment by fawce** [comment]
Closing as a duplicate of #2603
